seqr/views/utils/terra_api_utils_tests.py

```diff
@@ -179,28 +178,55 @@ def test_user_get_workspace_access_level(self, mock_time, mock_redis, mock_logge
         self.assertEqual(len(mock_logger.method_calls), 1)
         responses.assert_call_count(url, 1)
 
         mock_logger.reset_mock()
-        responses.replace(responses.GET, url, status = 404)
+        responses.replace(responses.GET, url, status=404)
         permission = user_get_workspace_access_level(user, 'my-seqr-billing', 'my-seqr-workspace')
         self.assertDictEqual(permission, {})
-        mock_logger.warning.assert_called_with(
-            'test_user called Terra API: GET /api/workspaces/my-seqr-billing/my-seqr-workspace?fields=accessLevel got status 404 with reason: Not Found')
-        self.assertEqual(len(mock_logger.method_calls), 1)
         responses.assert_call_count(url, 2)
 
         # Test cache hit
         mock_redis.return_value.get.return_value = '{"accessLevel": "READER"}'
         permission = user_get_workspace_access_level(user, 'my-seqr-billing', 'my-seqr-workspace')
         self.assertDictEqual(permission, {"accessLevel": "READER"})
         mock_redis.return_value.get.assert_called_with(
-            'terra_req__test_user__api/workspaces/my-seqr-billing/my-seqr-workspace?fields=accessLevel')
+            'terra_req__test_user__api/workspaces/my-seqr-billing/my-seqr-workspace?fields=accessLevel,canShare')
         responses.assert_call_count(url, 2)  # No API called since the call_count is kept unchanged.
+
+    @responses.activate
+    @mock.patch('seqr.views.utils.terra_api_utils.SERVICE_ACCOUNT_FOR_ANVIL', TEST_SERVICE_ACCOUNT)
+    @mock.patch('seqr.utils.redis_utils.redis.StrictRedis')
+    @mock.patch('seqr.views.utils.terra_api_utils.time')
+    def test_add_service_account(self, mock_time, mock_redis):
+        mock_redis.return_value.get.return_value = None
+
+        user = User.objects.get(username='test_user')
+        responses.add(responses.POST, GOOGLE_API_TOKEN_URL, status=200, body=GOOGLE_TOKEN_RESULT)
+        mock_time.time.return_value = AUTH_EXTRA_DATA['auth_time'] + 10
+
+        url = '{}api/workspaces/my-seqr-billing/my-seqr-workspace/acl'.format(TEST_TERRA_API_ROOT_URL)
+        responses.add(responses.GET, url, status=200, body='{{"acl": {{"{}": {{"accessLevel": "READER","canCompute": false,"canShare": false,"pending": false}} }} }}'.format(TEST_SERVICE_ACCOUNT))
+        r = add_service_account(user, 'my-seqr-billing', 'my-seqr-workspace')
+        self.assertTrue(r)
+        self.assertEqual(responses.calls[0].request.url, url)
+
+        responses.replace(responses.GET, url, status=200, body='{"acl": {}}')
+        responses.add(responses.PATCH, url, status=200, body='{{"usersUpdated": [{{"email": "{}" }}]}}'.format(TEST_SERVICE_ACCOUNT))
+        r = add_service_account(user, 'my-seqr-billing', 'my-seqr-workspace')
+        self.assertTrue(r)
+        self.assertEqual(responses.calls[1].request.url, url)
+        self.assertEqual(responses.calls[2].request.url, url)
+        self.assertEqual(responses.calls[2].request.method, responses.PATCH)
```

Comment:
why are there 2 api calls to the same url? If they are supposed to have different methods, can you explicitly test the method for both and not just for the second call?

Reply:
Same URL and different methods. I can add a test for the other call's method.

Reply:
you should test for the PATCH call that the body it was called with was correct